roaring.go

```diff
@@ -53,6 +53,133 @@ func (rb *Bitmap) ToBytes() ([]byte, error) {
 	return rb.highlowcontainer.toBytes()
 }
 
+const wordSize = uint64(64)
+const log2WordSize = uint64(6)
+const capacity = ^uint64(0)
+const bitmapContainerSize = (1 << 16) / 64 // bitmap size in words
+
+// DenseSize returns the size of the bitmap when stored as a dense bitmap.
+func (rb *Bitmap) DenseSize() int {
+	if rb.highlowcontainer.size() == 0 {
+		return 0
+	}
+
+	maximum := 1 + uint64(rb.Maximum())
+	if maximum > (capacity - wordSize + 1) {
+		return int(capacity >> log2WordSize)
+	}
+
+	return int((maximum + (wordSize - 1)) >> log2WordSize)
+}
+
+// ToDense returns a slice of uint64s representing the bitmap as a dense bitmap.
+// Useful to convert a roaring bitmap to a format that can be used by other libraries
+// like https://github.com/bits-and-blooms/bitset or https://github.com/kelindar/bitmap
+func (rb *Bitmap) ToDense() []uint64 {
+	sz := rb.DenseSize()
+	if sz == 0 {
+		return nil
+	}
+
+	bitmap := make([]uint64, sz)
+	rb.WriteDenseTo(bitmap)
+	return bitmap
+}
+
+// FromDense creates a bitmap from a slice of uint64s representing the bitmap as a dense bitmap.
+// Useful to convert bitmaps from libraries like https://github.com/bits-and-blooms/bitset or
+// https://github.com/kelindar/bitmap into roaring bitmaps fast and with convenience.
+func FromDense(bitmap []uint64) *Bitmap {
+	sz := (len(bitmap) + bitmapContainerSize - 1) / bitmapContainerSize // round up
+	rb := &Bitmap{
+		highlowcontainer: roaringArray{
+			containers:      make([]container, 0, sz),
+			keys:            make([]uint16, 0, sz),
+			needCopyOnWrite: make([]bool, 0, sz),
+		},
+	}
+	rb.FromDense(bitmap)
+	return rb
+}
+
+// FromDense unmarshalls from a slice of uint64s representing the bitmap as a dense bitmap.
+// Useful to convert bitmaps from libraries like https://github.com/bits-and-blooms/bitset or
+// https://github.com/kelindar/bitmap into roaring bitmaps fast and with convenience.
+// Callers are responsible for ensuring that the bitmap is empty before calling this function.
+func (rb *Bitmap) FromDense(bitmap []uint64) {
+	if len(bitmap) == 0 {
+		return
+	}
+
+	var k uint16
+	const size = bitmapContainerSize
+	cow := true
+
+	for len(bitmap) > 0 {
+		if len(bitmap) < size {
+			bm := make([]uint64, size)
+			copy(bm, bitmap)
+			bitmap = bm
+			cow = false
+		}
+
+		words := bitmap[:size]
+		if count := int(popcntSlice(words)); count > 0 {
+			c := &bitmapContainer{
+				bitmap:      words,
+				cardinality: count,
+			}
+			rb.highlowcontainer.appendContainer(k, c, cow)
```

Comment:
That won't work because there is a constraint. We do not allow bitmap containers with `arrayDefaultMaxSize` elements or fewer. Thankfully, you can call `toEfficientContainer()`.

Note that when we want to pick a bitmap container, you will need to allocate a new slice and copy because doing `words := bitmap[:size]` is problematic since it points at the bitmap slice.


https://github.com/RoaringBitmap/roaring/blob/c79926f24cdcf0dc5fa59f688c308f0045e3b182/bitmapcontainer.go#L1089

Reply:
> That won't work because there is a constraint. We do not allow bitmap containers with arrayDefaultMaxSize elements or fewer. Thankfully, you can call toEfficientContainer().

Ah, makes sense!

> Note that when we want to pick a bitmap container, you will need to allocate a new slice and copy because doing words := bitmap[:size] is problematic since it points at the bitmap slice.

I assumed that passing the copy on write argument as true would allow us to not copy. Can you clarify what that argument is for if not for that?

Reply:
Pushed d82c20ddaf065bbf7427ab7a27a7bdd04e5cdaf0.

Opted for not using `toEfficientContainer()` for performance reasons. Callers can still call `RunOptimize()` if they want afterwards.

    ```
    name \ time/op                        before         with-efficient-container  no-to-efficient-container
    FromDense/bitmap-4097-10                1.45µs ± 1%               2.15µs ± 0%                1.47µs ± 2%
    FromDense/run-4098-10                   1.04µs ± 4%               3.00µs ± 0%                1.04µs ± 2%
    FromDense/array-4096-10                 1.45µs ± 1%               6.47µs ± 8%                5.46µs ± 1%
    FromDense/bitmaps-and-runs-300000-10    3.53µs ± 0%              13.53µs ± 1%                3.57µs ± 1%

    name \ speed                          before         with-efficient-container  no-to-efficient-container
    FromDense/bitmap-4097-10              6.35GB/s ± 1%             4.28GB/s ± 0%              6.27GB/s ± 2%
    FromDense/run-4098-10                  989MB/s ± 4%              345MB/s ± 0%               990MB/s ± 2%
    FromDense/array-4096-10               6.36GB/s ± 1%             1.43GB/s ± 8%              1.69GB/s ± 1%
    FromDense/bitmaps-and-runs-300000-10  15.9GB/s ± 0%              4.2GB/s ± 1%              15.8GB/s ± 1%

    name \ alloc/op                       before         with-efficient-container  no-to-efficient-container
    FromDense/bitmap-4097-10                8.22kB ± 0%               8.22kB ± 0%                8.22kB ± 0%
    FromDense/run-4098-10                   8.22kB ± 0%               8.26kB ± 0%                8.22kB ± 0%
    FromDense/array-4096-10                 8.22kB ± 0%              16.44kB ± 0%               16.41kB ± 0%
    FromDense/bitmaps-and-runs-300000-10    8.35kB ± 0%               8.49kB ± 0%                8.35kB ± 0%

    name \ allocs/op                      before         with-efficient-container  no-to-efficient-container
    FromDense/bitmap-4097-10                  2.00 ± 0%                 2.00 ± 0%                  2.00 ± 0%
    FromDense/run-4098-10                     2.00 ± 0%                 4.00 ± 0%                  2.00 ± 0%
    FromDense/array-4096-10                   2.00 ± 0%                 4.00 ± 0%                  3.00 ± 0%
    FromDense/bitmaps-and-runs-300000-10      6.00 ± 0%                16.00 ± 0%                  6.00 ± 0%


Holding off on introducing the copy until I understand the `mustCopyOnWrite` semantics.